File: modules/embeddings/exercises/02_cosine_similarity/solution/cpp/main.cpp
#include <cmath>
#include <iomanip>
#include <iostream>
#include <vector>

static double cosine(const std::vector<double>& a, const std::vector<double>& b) {
    double dot = 0, na = 0, nb = 0;
    for (size_t i = 0; i < a.size(); ++i) {
        dot += a[i] * b[i];
        na += a[i] * a[i];
        nb += b[i] * b[i];
    }
    return dot / (std::sqrt(na) * std::sqrt(nb));
}

int main() {
    std::cout << "SIMILARITY:" << std::fixed << std::setprecision(1) << cosine({1, 0, 0}, {1, 0, 0}) << "\n";
    return 0;
}

File: modules/embeddings/exercises/02_cosine_similarity/solution/python/main.py
import math
def cosine(a,b):
    dot=sum(x*y for x,y in zip(a,b))
    na=math.sqrt(sum(x*x for x in a)); nb=math.sqrt(sum(x*x for x in b))
    return dot/(na*nb)
def main():
    print(f"SIMILARITY:{cosine([1,0,0],[1,0,0]):.1f}")
if __name__=="__main__": main()
